pkg/chain/ethereum/tbtc.go

```diff
@@ -370,214 +674,195 @@ func (tc *TbtcChain) GetDKGState() (tbtc.DKGState, error) {
-) subscription.EventSubscription {
-	return tc.mockWalletRegistry.OnSignatureRequested(handler)
+	return computeDkgResultHash(
+		tc.chainID,
+		unprefixedGroupPublicKey,
+		tecdsaDkgResult.MisbehavedMembersIndexes(),
+		big.NewInt(int64(startBlock)),
+	)
 }
 
-// TODO: Temporary mock that simulates the behavior of the WalletRegistry
-//
-//	contract. Should be removed eventually.
-type mockWalletRegistry struct {
-	blockCounter chain.BlockCounter
-
-	dkgResultSubmissionHandlersMutex sync.Mutex
-	dkgResultSubmissionHandlers      map[int]func(submission *tbtc.DKGResultSubmittedEvent)
-
-	currentDkgMutex      sync.RWMutex
-	currentDkgStartBlock *big.Int
-
-	activeWalletMutex         sync.RWMutex
-	activeWallet              []byte
-	activeWalletOperableBlock *big.Int
-}
+// computeDkgResultHash computes the keccak256 hash for the given DKG result
+// parameters.
+func computeDkgResultHash(
+	chainID *big.Int,
+	groupPublicKey []byte,
+	misbehavedMembersIndexes []group.MemberIndex,
+	startBlock *big.Int,
+) (dkg.ResultHash, error) {
+	publicKeySize := 64
 
-func newMockWalletRegistry(blockCounter chain.BlockCounter) *mockWalletRegistry {
-	return &mockWalletRegistry{
-		blockCounter: blockCounter,
-		dkgResultSubmissionHandlers: make(
-			map[int]func(submission *tbtc.DKGResultSubmittedEvent),
-		),
+	if len(groupPublicKey) != publicKeySize {
+		return dkg.ResultHash{}, fmt.Errorf("wrong group public key length")
 	}
-}
 
-func (mwr *mockWalletRegistry) OnDKGStarted(
-	handler func(event *tbtc.DKGStartedEvent),
-) subscription.EventSubscription {
-	ctx, cancelCtx := context.WithCancel(context.Background())
-	blocksChan := mwr.blockCounter.WatchBlocks(ctx)
+	uint256Type, err := abi.NewType("uint256", "uint256", nil)
+	if err != nil {
+		return dkg.ResultHash{}, err
+	}
+	bytesType, err := abi.NewType("bytes", "bytes", nil)
+	if err != nil {
+		return dkg.ResultHash{}, err
+	}
+	uint8SliceType, err := abi.NewType("uint8[]", "uint8[]", nil)
+	if err != nil {
+		return dkg.ResultHash{}, err
+	}
 
-	go func() {
-		for {
-			select {
-			case block := <-blocksChan:
-				// Generate an event every 1000th block starting from block 250.
-				// The shift is done in order to avoid overlapping with beacon
-				// DKG test loop.
-				shift := uint64(250)
-				if block >= shift && (block-shift)%1000 == 0 {
-					// The seed is keccak256(block).
-					blockBytes := make([]byte, 8)
-					binary.BigEndian.PutUint64(blockBytes, block)
-					seedBytes := crypto.Keccak256(blockBytes)
-					seed := new(big.Int).SetBytes(seedBytes)
-
-					mwr.currentDkgStartBlock = big.NewInt(int64(block))
-
-					go handler(&tbtc.DKGStartedEvent{
-						Seed:        seed,
-						BlockNumber: block,
-					})
-				}
-			case <-ctx.Done():
-				return
-			}
-		}
-	}()
+	bytes, err := abi.Arguments{
+		{Type: uint256Type},
+		{Type: bytesType},
+		{Type: uint8SliceType},
+		{Type: uint256Type},
+	}.Pack(
+		chainID,
+		groupPublicKey,
+		misbehavedMembersIndexes,
+		startBlock,
+	)
```

Comment:
I am looking at `EcdsaDkg.submitResult` and the hash is computed as:
https://github.com/keep-network/keep-core/blob/089f9b549647feda54d4f86a86480dc4514736b6/solidity/ecdsa/contracts/libraries/EcdsaDkg.sol#L268

and there are more fields in `Result` structure compared to the ones included here:
https://github.com/keep-network/keep-core/blob/089f9b549647feda54d4f86a86480dc4514736b6/solidity/ecdsa/contracts/libraries/EcdsaDkg.sol#L86-L118

Did we miss something?

Reply:
Please also see [this](https://github.com/keep-network/keep-core/pull/3427#discussion_r1053172506) comment.

Reply:
Well, we actually have two sorts of hashes. 

The first sort is used when members sign the result to show their support and mutually verify their signatures before attempting to submit the result. This hash is also used during the result validation to make sure group members signed the right thing. See https://github.com/keep-network/keep-core/blob/089f9b549647feda54d4f86a86480dc4514736b6/solidity/ecdsa/contracts/EcdsaDkgValidator.sol#L221 This is the hash the client needs to compute by itself.

The second type of hash is the hash you mentioned here. This one is used internally by the contract to store the result in a gas-efficient way. The client does not need to compute it though that hash is populated in DKG-related events that are emitted after submission.

I should make it clear in the code and use better naming. I'll try to improve that.

Reply:
See 07d4b585